### ISSUE TITLE
How does yolov8 use pycocotools to validate the coco dataset, I need to get its performance in AP 'l', 's', 'm' and AP75

### ISSUE THREAD

**Issue by tangruinenu** [ultralytics/ultralytics#1745]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How does yolov8 use pycocotools to validate the coco dataset, I need to get its performance in AP 'l', 's', 'm' and AP75

**Comment by glenn-jocher** [comment]
@tangruinenu pycocotools is used automatically when COCO is the dataset, i.e.

```
yolo val data=coco.yaml
```

**Comment by tangruinenu** [comment]
> @tangruinenu pycocotools is used automatically when COCO is the dataset, i.e.
> 
> ```
> yolo val data=coco.yaml
> ```
I did the verification according to what you said, YOLOv8 did not actively use pycocotools to evaluate the COCO validation set.

<img width="1058" alt="截屏2023-04-02 下午3 07 34" src="https://user-images.githubusercontent.com/100890412/229337995-8b3a3bcf-d4cb-4b1c-a4a1-587e1b2b419f.png">
<img width="1327" alt="截屏2023-04-02 下午12 40 19" src="https://user-images.githubusercontent.com/100890412/229338012-d91440e7-fc43-40c4-a8f2-3fb232c2eeed.png">

**Comment by tangruinenu** [comment]
> @tangruinenu pycocotools is used automatically when COCO is the dataset, i.e.
> 
> ```
> yolo val data=coco.yaml
> ```
I did the verification according to what you said, YOLOv8 did not actively use pycocotools to evaluate the COCO validation set.
<img width="1075" alt="截屏2023-04-02 下午3 08 50" src="https://user-images.githubusercontent.com/100890412/229338040-15e97b0a-badf-42c3-a85a-e83677e23d02.png">
<img width="769" alt="截屏2023-04-02 下午3 07 53" src="https://user-images.githubusercontent.com/100890412/229338043-fd9aa042-7662-443f-ad42-1055bdcf8255.png">

**Comment by glenn-jocher** [comment]
@tangruinenu I apologize for the confusion in my previous response. It looks like you are right that YOLOv8 does not currently use pycocotools for evaluation. The evaluation is done using a script called `test.py`. Currently, the test script outputs mAP and AP50 values only. If you need AP `l`, `m`, and `s` values, you may need to modify this script to calculate them separately. Alternatively, you could use pycocotools directly to compute AP values. Sorry for any confusion caused.

**Comment by glenn-jocher** [comment]
@tangruinenu that's very strange. I can't account for the reason pycocotools is not being used locally on your machine. You may need to install it first, i.e. pip install pycocotools before running a COCO evaluation. If that doesn't work, I'd recommend running one of our verified environments below, where pycocotools and all dependencies are already installed correctly.

## Environments

YOLOv8 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
- **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

**Comment by tangruinenu** [comment]
> 

I already installed this, because I was using YOLOX, and his validation was always using pycocotools

**Comment by tangruinenu** [comment]
Also, I have tested on Gpus on various devices and there is no way to automatically use pycocotools. So let me know if there are methods that are successfully called manually in yolov8. I tried it myself and it was almost zero.

**Comment by tangruinenu** [comment]
> @tangruinenu that's very strange. I can't account for the reason pycocotools is not being used locally on your machine. You may need to install it first, i.e. pip install pycocotools before running a COCO evaluation. If that doesn't work, I'd recommend running one of our verified environments below, where pycocotools and all dependencies are already installed correctly.
> 
> ## Environments
> YOLOv8 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
> 
> * **Notebooks** with free GPU: [![Run on Gradient](https://camo.githubusercontent.com/c9931a1689c37ab786edd3e1e5f59b9a6f7d097628c4689ce2432563ef884524/68747470733a2f2f6173736574732e706170657273706163652e696f2f696d672f6772616469656e742d62616467652e737667)](https://console.paperspace.com/github/ultralytics/ultralytics) [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/ultralytics/yolov8)
> * **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
> * **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
> * **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) [![Docker Pulls](https://camo.githubusercontent.com/7401823bed1473d6db1ff60b933d51cd1534c448488f06c446c322227d31d966/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f756c7472616c79746963733f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/ultralytics)

Also, I have tested on Gpus on various devices and there is no way to automatically use pycocotools. So let me know if there are methods that are successfully called manually in yolov8. I tried it myself and it was almost zero.

**Comment by tangruinenu** [comment]
> @tangruinenu that's very strange. I can't account for the reason pycocotools is not being used locally on your machine. You may need to install it first, i.e. pip install pycocotools before running a COCO evaluation. If that doesn't work, I'd recommend running one of our verified environments below, where pycocotools and all dependencies are already installed correctly.
> 
> ## Environments
> YOLOv8 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
> 
> * **Notebooks** with free GPU: [![Run on Gradient](https://camo.githubusercontent.com/c9931a1689c37ab786edd3e1e5f59b9a6f7d097628c4689ce2432563ef884524/68747470733a2f2f6173736574732e706170657273706163652e696f2f696d672f6772616469656e742d62616467652e737667)](https://console.paperspace.com/github/ultralytics/ultralytics) [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/ultralytics/yolov8)
> * **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
> * **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
> * **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) [![Docker Pulls](https://camo.githubusercontent.com/7401823bed1473d6db1ff60b933d51cd1534c448488f06c446c322227d31d966/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f756c7472616c79746963733f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/ultralytics)

I already installed this, because I was using YOLOX, and his validation was always using pycocotools

**Comment by glenn-jocher** [comment]
@tangruinenu I apologize for the confusion, it seems I provided an incorrect initial response. Upon further testing, it appears that YOLOv8 indeed does not currently use pycocotools in its validation script. We are aware of the issue and it is on our roadmap to support it in the future. In the meantime, I can suggest using other popular COCO evaluation repositories like cocoapi or mAP for evaluating YOLOv8's performance on the COCO dataset. 

Thank you for bringing this to our attention, and please let me know if there's anything else I can assist you with.
